src/Cassandra/Requests/RequestHandler.cs

```diff
@@ -247,54 +240,95 @@ private Host GetNextHost()
+        private bool TryValidateHost(Host host, out ValidHost validHost)
+        {
+            var distance = Cluster.RetrieveDistance(host, Policies.LoadBalancingPolicy);
+            validHost = ValidHost.New(host, distance);
+            return validHost != null;
+        }
+
+        /// <inheritdoc />
+        public async Task<IConnection> GetNextConnectionAsync(Dictionary<IPEndPoint, Exception> triedHosts)
+        {
+            Host host;
+            // While there is an available host
+            while ((host = GetNextHost()) != null)
+            {
+                var c = await ValidateHostAndGetConnectionAsync(host, triedHosts).ConfigureAwait(false);
                 if (c == null)
                 {
                     continue;
                 }
                 return c;
             }
+
             throw new NoHostAvailableException(triedHosts);
         }
 
+        /// <inheritdoc />
+        public async Task<IConnection> ValidateHostAndGetConnectionAsync(Host host, Dictionary<IPEndPoint, Exception> triedHosts)
+        {
+            if (_session.IsDisposed)
+            {
+                throw new NoHostAvailableException(triedHosts);
+            }
+
+            triedHosts[host.Address] = null;
+            if (!TryValidateHost(host, out var validHost))
+            {
+                return null;
+            }
+            
+            var c = await GetConnectionToValidHostAsync(validHost, triedHosts).ConfigureAwait(false);
+            return c;
+        }
+
+        /// <inheritdoc />
+        public Task<IConnection> GetConnectionToValidHostAsync(ValidHost validHost, IDictionary<IPEndPoint, Exception> triedHosts)
+        {
+            return RequestHandler.GetConnectionFromHostAsync(validHost.Host, validHost.Distance, _session, triedHosts);
+        }
+
         /// <summary>
         /// Gets a connection from a host or null if its not possible, filling the triedHosts map with the failures.
         /// </summary>
-        internal static async Task<Connection> GetConnectionFromHost(Host host, HostDistance distance, Session session,
-                                                                     IDictionary<IPEndPoint, Exception> triedHosts)
+        /// <param name="host">Host to which a connection will be obtained.</param>
+        /// <param name="distance">Output parameter that will contain the <see cref="HostDistance"/> associated with
+        /// <paramref name="host"/>. It is retrieved from the current <see cref="ILoadBalancingPolicy"/>.</param>
+        /// <param name="session">Session from where a connection will be obtained (or created).</param>
+        /// <param name="triedHosts">Hosts for which there were attempts to connect and send the request.</param>
+        /// <exception cref="InvalidQueryException">When the keyspace is not valid</exception>
+        /// <exception cref="NoHostAvailableException"></exception>
```

Comment:
NIT: this exception is not thrown by this method, it stood out to me its weird to get `NoHostAvailableException` when the host is the parameter.

Reply:
nice catch, fixed